src/OrchardCore/OrchardCore.Abstractions/Shell/Builders/ShellContext.cs

```diff
@@ -10,19 +13,19 @@ namespace OrchardCore.Environment.Shell.Builders
     /// The shell context represents the shell's state that is kept alive
     /// for the whole life of the application
     /// </summary>
-    public class ShellContext : IDisposable
+    public class ShellContext : IDisposable, IAsyncDisposable
     {
         private bool _disposed;
         private List<WeakReference<ShellContext>> _dependents;
-        private readonly object _synLock = new object();
+        private readonly SemaphoreSlim _semaphore = new(1);
 
         internal volatile int _refCount;
         internal volatile int _terminated;
         internal bool _released;
 
         public ShellSettings Settings { get; set; }
         public ShellBlueprint Blueprint { get; set; }
-        public IServiceProvider ServiceProvider { get; set; }
+        public ServiceProvider ServiceProvider { get; set; }
```

Comment:
Required? That's a breaking change, and that's weird to have to return the actual class.

Reply:
Or keep the interface using a public accessor and make the class internal

Reply:
I thought that changing `ServiceProvider` to `IServiceProvider` would be a breaking change but not the opposite. It allows to not have to do a casting to `IDisposable` or `IDisposableAsync` before calling `Dispose()` or `DisposeAsync()`. For info we only set it when we call `_shellContainerFactory.CreateContainerAsync()`.

But no problem to keep it as a `IServiceProvider` if you prefer.

Reply:
Okay I reverted it to `IServiceProvider` and re-used the castings.